pkg/operator/controller/machineconfig/nodes.go

```diff
@@ -25,109 +25,128 @@ import (
 
-		updNodeCount++
+	r.Log.V(1).Info("nodeobservability role is not removed on all the nodes with worker role", "Nodes", requiredNodeCount, "UpdatedNodeCount", nodeCount)
+	return fmt.Errorf("failed to ensure removal of labels, expected updates on %d nodes but got %d", requiredNodeCount, nodeCount)
+}
+
+// listNodes returns the list of nodes matching the labels
+func (r *MachineConfigReconciler) listNodes(ctx context.Context, matchLabels map[string]string) (*corev1.NodeList, error) {
+	listOpts := []client.ListOption{
+		client.MatchingLabels(matchLabels),
 	}
 
-	if updNodeCount == len(nodeList.Items) {
-		r.Log.V(1).Info("Successfully removed nodeobservability role from nodes with worker role", "NodeCount", updNodeCount)
+	nodeList := &corev1.NodeList{}
+	if err := r.ClientList(ctx, nodeList, listOpts...); err != nil {
+		return nil, err
 	}
-	return updNodeCount, nil
+
+	return nodeList, nil
 }
 
-// listNoObsLabeledNodes returns the list of nodes having NodeObservability role label
-func (r *MachineConfigReconciler) listNoObsLabeledNodes(ctx context.Context, nodeList *corev1.NodeList) error {
-	nodeLabel := map[string]string{
-		NodeObservabilityNodeRoleLabelName: Empty,
+// ensureProfConfEnabled makes sure all the configuration needed to enable the
+// CRI-O profiling is applied.
+func (r *MachineConfigReconciler) ensureProfConfEnabled(ctx context.Context, nomc *v1alpha2.NodeObservabilityMachineConfig) error {
+
+	nodeList, err := r.listNodes(ctx, nomc.Spec.NodeSelector)
+	if err != nil {
+		return err
+	}
+
+	if len(nodeList.Items) == 0 {
+		r.Log.V(1).Info("No nodes matching the given selector were found", "NodeSelector", nomc.Spec.NodeSelector)
+		return nil
+	}
+
+	err = r.ensureReqNodeLabelExists(ctx, nodeList)
+	if err != nil {
+		return fmt.Errorf("failed to ensure nodes are labelled: %w", err)
 	}
 
-	if err := r.listNodes(ctx, nodeList, nodeLabel); err != nil {
-		return fmt.Errorf("failed to get the list of nodes labeled nodeobservability: %w", err)
+	if err := r.enableCrioProf(ctx, nomc); err != nil {
+		return fmt.Errorf("failed to enable CRIO mc due to: %w", err)
+	}
+
+	if err := r.createProfMCP(ctx, nomc); err != nil {
+		return fmt.Errorf("failed to create profiling mcp due to: %w", err)
 	}
 
 	return nil
 }
 
-// listNodes returns the list of nodes matching the labels
-func (r *MachineConfigReconciler) listNodes(ctx context.Context, nodeList *corev1.NodeList, matchLabels map[string]string) error {
-	listOpts := []client.ListOption{
-		client.MatchingLabels(matchLabels),
+// ensureProfConfDisabled disables the profiling on the requested nodes by removing the nodeobservability label.
+func (r *MachineConfigReconciler) ensureProfConfDisabled(ctx context.Context, nomc *v1alpha2.NodeObservabilityMachineConfig) error {
+
+	nodeList, err := r.listNodes(ctx, map[string]string{NodeObservabilityNodeRoleLabelName: empty})
+	if err != nil {
+		return fmt.Errorf("failed to get the list of nodes labelled nodeobservability: %w", err)
```

Comment:
According to previous implementation, we seem to have hard coded the label selector used while disabling CRI-O profiling whereas during enabling we seem to taking it as user input. This doesn't feel right, I think we should enable and disable using the user given labels only. WDYT?

Reply:
At the time of the disabling those are the same label selector as we add nob role label only to the nodes selected by the user provided label selector. I believe we only use the user provided one when we label nodes, in all the other cases (here and in nob `MCP`) we use the created label.